client/src/lib/metrics/index.tsx

```diff
@@ -0,0 +1,73 @@
+import React from 'react';
+import PostHog, {PostHogProvider} from 'posthog-react-native';
+import {POSTHOG_API_KEY} from 'config';
+import {CoreProperties, Events, UserProperties} from './types';
+import {
+  UnionToIntersection,
+  Values,
+} from '../../../../shared/src/types/UtilityTypes';
+
+const DEFAULT_CONSENT = true;
+
+let postHog: PostHog | undefined;
+let initPromise: Promise<PostHog> | undefined;
+
+export const init = async () => {
+  if (!POSTHOG_API_KEY) {
+    return initPromise;
+  }
+
+  if (!initPromise) {
+    postHog = await (initPromise = PostHog.initAsync(POSTHOG_API_KEY, {
+      host: 'https://eu.posthog.com',
+    }));
+
+    setConsent(DEFAULT_CONSENT);
+  }
+
+  return initPromise;
+};
+
+export const setConsent = async (haveConsent: boolean) =>
+  haveConsent ? (await init())?.optIn() : (await init())?.optOut();
+
+export const MetricsProvider: React.FC<{children: React.ReactNode}> = ({
+  children,
+}) => <PostHogProvider client={postHog}>{children}</PostHogProvider>;
+
+type AnyUserProperty = Partial<UserProperties>;
+type AnyCoreProperty = Partial<CoreProperties>;
+
+// Posthog event user properties https://posthog.com/docs/integrate/user-properties
+type EventUserProperties = {
+  $set?: AnyUserProperty;
+  $set_once?: AnyUserProperty;
+  $unset?: [keyof UserProperties];
+};
+
+// Generate all possible allowed event names and properties as logEvent overloads
+type LogEventOverloads = UnionToIntersection<
```

Comment:
black magic right there hahaha
nice to get typed stuff on metrics

Reply:
Thanks! 🤗